src/ngx_http_lua_output.c

```diff
@@ -802,4 +802,105 @@ ngx_http_lua_flush_cleanup(void *data)
     ctx->flushing_coros--;
 }
 
+
+#ifndef NGX_LUA_NO_FFI_API
+int
+ngx_http_lua_ffi_write(ngx_http_request_t *r, const char *buf, size_t offset,
+                       size_t len, char **err)
+{
+    ngx_http_lua_ctx_t          *ctx;
+    ngx_uint_t                   size;
+    ngx_buf_t                   *b;
+    ngx_chain_t                 *cl;
+    ngx_int_t                    rc;
+    const char                  *str_sub;
+
+    if (r == NULL) {
+        *err = "no request object found";
+        return NGX_ERROR;
+    }
+
+    if (buf == NULL) {
+        *err = "no buffer found";
+        return NGX_ERROR;
+    }
+
+    size = ngx_strlen(buf);
```

Comment:
@agentzh I think there is a misunderstanding about the API in output.lua (same applies for ffi.cdef):
    
    ffi_write(lua_string, offset, len)
    
    lua_string: the lua string
    offset: the offset in the lua string
    len: the length to print starting from the offset
    
We are missing the length of the lua string. We should probably have the following API function instead:
     
     ffi_write(lua_string, len, from, to)
     
     lua_string: the lua string to write
     len: the length of the lua string
     from: the index from which we start to write
     to: the index to which we stop to write
     
or 
     
      ffi_write(lua_string, lua_string_len, offset, len)
      
      lua_string: the lua string
      lua_string_len: the length of the lua string
      offset: the offset in the lua string
      len: the length to print starting from the offset